pkg/runtime/kubernetes/runtime.go

```diff
@@ -17,45 +17,53 @@ package kubernetes
 import (
 	"context"
 	"fmt"
-	"strings"
-	"sync"
-
-	"github.com/sealerio/sealer/pkg/registry"
-	"github.com/sealerio/sealer/pkg/runtime/kubernetes/kubeadm"
-	"github.com/sealerio/sealer/utils"
-	versionUtils "github.com/sealerio/sealer/utils/version"
-
 	"net"
 	"path/filepath"
+	"strings"
+	"sync"
 	"time"
 
 	"github.com/sealerio/sealer/common"
+	"github.com/sealerio/sealer/pkg/client/k8s"
+	"github.com/sealerio/sealer/pkg/registry"
 	"github.com/sealerio/sealer/pkg/runtime"
+	"github.com/sealerio/sealer/pkg/runtime/kubernetes/kubeadm"
 	v2 "github.com/sealerio/sealer/types/api/v2"
+	"github.com/sealerio/sealer/utils"
 	"github.com/sealerio/sealer/utils/platform"
 	"github.com/sealerio/sealer/utils/ssh"
 	strUtils "github.com/sealerio/sealer/utils/strings"
-	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2"
+	versionUtils "github.com/sealerio/sealer/utils/version"
 
 	"github.com/sirupsen/logrus"
 	"golang.org/x/sync/errgroup"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2"
 )
 
 type Config struct {
-	Vlog      int
-	VIP       string
-	RegConfig *registry.Config
 	// Clusterfile: the absolute path, we need to read kubeadm config from Clusterfile
 	ClusterFileKubeConfig *kubeadm.KubeadmConfig
-	APIServerDomain       string
 }
 
 // Runtime struct is the runtime interface for kubernetes
 type Runtime struct {
 	*sync.Mutex
+	// TODO: remove field cluster from runtime pkg
+	// just make runtime to use essential data from cluster, rather than the whole cluster scope.
 	cluster *v2.Cluster
+
+	// The KubeadmConfig used to setup the final cluster.
+	// Its data is from KubeadmConfig input from Clusterfile and KubeadmConfig in ClusterImage.
 	*kubeadm.KubeadmConfig
-	*Config
+	// *Config
```

Comment:
Feel confused about the KubeadmConfigFromClusterfile. I notice that "*kubeadm.KubeadmConfig" above is used to setup the final cluster, but what is KubeadmConfigFromClusterfile capable of?

Reply:
I think it is because that we lack the design of how many ways we allow to input KubeadmConfig to a cluster instance. If the sealer policy allows to input KubeadmConfig from Clusterfile, then `KubeadmConfigFromClusterfile` is useful.

I wish to make it strictly enough when we are evolving, otherwise it may be out of control. Still be design problem, I think. @justadogistaken